relayer/src/supervisor/client_state_filter.rs

```diff
@@ -0,0 +1,274 @@
+        trace!(
+            "[client filter] deciding if to relay on {:?} hosted chain {}",
+            obj.dst_client_id,
+            obj.dst_chain_id
+        );
+        let client_state = chain.query_client_state(&obj.dst_client_id, Height::zero())?;
+        Ok(self.control_client(&obj.dst_client_id, &client_state))
+    }
+
+    pub fn control_conn_object(
+        &mut self,
+        registry: &mut Registry,
+        obj: &object::Connection,
+    ) -> Result<Permission, BoxError> {
+        let identifier =
+            CacheKey::Connection(obj.src_chain_id.clone(), obj.src_connection_id.clone());
+        trace!("controlling permissions for {:?}", identifier);
+
+        // Return if cache hit
+        if let Some(p) = self.permission_cache.get(&identifier) {
+            trace!("cache hit {:?} for {:?}", p, identifier);
+            return Ok(*p);
+        }
+
+        let src_chain = registry.get_or_spawn(&obj.src_chain_id)?;
+        trace!(
+            "[filter] deciding if to relay on {:?} hosted on chain {}",
+            obj,
+            obj.src_chain_id
+        );
+        let connection_end = src_chain.query_connection(&obj.src_connection_id, Height::zero())?;
+        let client_state =
+            src_chain.query_client_state(&connection_end.client_id(), Height::zero())?;
+
+        self.control_connection_end_and_client(
+            registry,
+            &obj.src_chain_id,
+            &client_state,
+            &connection_end,
+            &obj.src_connection_id,
+        )
+    }
+
+    fn control_channel(
+        &mut self,
+        registry: &mut Registry,
+        chain_id: &ChainId,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+    ) -> Result<Permission, BoxError> {
+        let identifier = CacheKey::Channel(chain_id.clone(), port_id.clone(), channel_id.clone());
+        trace!("controlling permissions for {:?}", identifier);
+
+        // Return if cache hit
+        if let Some(p) = self.permission_cache.get(&identifier) {
+            trace!("cache hit {:?} for {:?}", p, identifier);
+            return Ok(*p);
+        }
+
+        let src_chain = registry.get_or_spawn(&chain_id)?;
+        let channel_end = src_chain.query_channel(&port_id, &channel_id, Height::zero())?;
+        let conn_id = channel_end.connection_hops.first().ok_or_else(|| {
+            Kind::InvalidConnectionHopsLength(1, channel_end.connection_hops().len())
+        })?;
+        let connection_end = src_chain.query_connection(conn_id, Height::zero())?;
+        let client_state =
+            src_chain.query_client_state(&connection_end.client_id(), Height::zero())?;
+
+        let permission = self.control_connection_end_and_client(
+            registry,
+            &chain_id,
+            &client_state,
+            &connection_end,
+            &conn_id,
+        )?;
+
+        let key = CacheKey::from((chain_id.clone(), port_id.clone(), channel_id.clone()));
```

Comment:
Do we really need these `From` instances? It obscures a bit which type of `CacheKey` will be created. How about doing this instead?

```suggestion
        let key = CacheKey::Channel(chain_id.clone(), port_id.clone(), channel_id.clone());
```

Reply:
Good point. I think I was just enthusiastic about it and didn't realize this simpler approach, nice!

3abdf1de